packages/SwingSet/src/liveslots/watchedPromises.js

```diff
@@ -0,0 +1,162 @@
+    }
+
+    p.then(
+      res => settle(res, true),
+      rej => settle(rej, false),
+    );
+  }
+
+  function loadWatchedPromiseTable() {
+    const deadPromisesRaw = syscall.vatstoreGet('deadPromises');
+    if (deadPromisesRaw) {
+      const deadPromises = new Set(deadPromisesRaw.split(','));
+      syscall.vatstoreDelete('deadPromises');
+      for (const [vpid, watches] of watchedPromiseTable.entries()) {
+        if (deadPromises.has(vpid)) {
+          watchedPromiseTable.delete(vpid);
+          for (const watch of watches) {
+            const [watcher, ...args] = watch;
+            if (watcher.onRejected) {
+              watcher.onRejected('vat upgraded', ...args);
+            }
+          }
+        } else {
+          const p = convertSlotToVal(vpid);
+          pseudoThen(p, vpid);
+        }
+      }
+    }
+  }
+
+  // defined promise watcher objects indexed by kindHandle
+  let promiseWatcherByKindTable;
+
+  function preparePromiseWatcherTables() {
+    let watcherTableID = syscall.vatstoreGet('watcherTableID');
+    if (watcherTableID) {
+      promiseWatcherByKindTable = convertSlotToVal(watcherTableID);
+    } else {
+      promiseWatcherByKindTable = makeScalarBigMapStore(
+        'promiseWatcherByKind',
+        { durable: true },
+      );
+      watcherTableID = convertValToSlot(promiseWatcherByKindTable);
+      syscall.vatstoreSet('watcherTableID', watcherTableID);
+      // artificially increment the table's refcount so it never gets GC'd
+      vrm.addReachableVref(watcherTableID);
+    }
+
+    let watchedPromiseTableID = syscall.vatstoreGet('watchedPromiseTableID');
+    if (watchedPromiseTableID) {
+      watchedPromiseTable = convertSlotToVal(watchedPromiseTableID);
+    } else {
+      watchedPromiseTable = makeScalarBigMapStore('watchedPromises', {
+        keySchema: M.string(),
+        durable: true,
+      });
+      watchedPromiseTableID = convertValToSlot(watchedPromiseTable);
+      syscall.vatstoreSet('watchedPromiseTableID', watchedPromiseTableID);
+      // artificially increment the table's refcount so it never gets GC'd
+      vrm.addReachableVref(watchedPromiseTableID);
+    }
+  }
+
+  function watchPromise(p, watcher, ...args) {
+    const watcherVref = convertValToSlot(watcher);
+    assert(watcherVref, 'invalid watcher');
+    const { virtual } = parseVatSlot(watcherVref);
+    assert(virtual, 'promise watcher must be a virtual object');
+    if (watcher.onFulfilled) {
+      assert.typeof(watcher.onFulfilled, 'function');
+    }
+    if (watcher.onRejected) {
+      assert.typeof(watcher.onRejected, 'function');
+    }
+
+    const vpid = convertValToSlot(p);
+    assert(vpid, 'invalid promise');
+    const { type } = parseVatSlot(vpid);
+    assert(type === 'promise', 'watchPromise only watches promises');
+    if (watchedPromiseTable.has(vpid)) {
+      const watches = watchedPromiseTable.get(vpid);
+      watchedPromiseTable.set(vpid, harden([...watches, [watcher, ...args]]));
```

Comment:
Out of curiosity, why not nest a durable scalar set instead of rewriting an array value?

Reply:
I thought about that, but since we expect the typical cardinality of this array to be small (in fact, usually 1), I think the overhead of a durable set would almost always dwarf the cost of rewriting the array.  Plus, the entry is the watcher plus its args, so it would end up being a set of copy arrays (even more expensive).  Either that or a map from a nonce or counter to the watcher plus args, and then we'd also have to keep track of the counter.  Anyway, a tarbaby of multiplying complications.

Reply:
I figured that was the case.